docs/design/2022-03-16-functional-dependency.md

```diff
@@ -0,0 +1,147 @@
+|     │ └─Selection_14             | 9.99     | cop[tikv] |               | eq(test.t1.c, 1), not(isnull(test.t1.a))                                                                                         |
+|     │   └─TableFullScan_13       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                                   |
+|     └─TableReader_18(Probe)      | 9990.00  | root      |               | data:Selection_17                                                                                                                |
+|       └─Selection_17             | 9990.00  | cop[tikv] |               | not(isnull(test.t2.a))                                                                                                           |
+|         └─TableFullScan_16       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                                   |
++----------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------------------+
+9 rows in set (0.00 sec)
+```
+
+Analysis:
+
+When building the FD through these logical operators, we are calling a function of ExtractFD recursively from the root node of the logical plan tree. When the call down to the leaf node (actually source table for both join side), for the inner side, the FD from the datasource of base table of t2 are nil because they have no basic key constraint. While for outer side, since there is a filter on the datasource of base table t1, we got FD as {} -> {t1.c}. Because t1.c can only be constant 1 behind the filter. Up to next level, the join condition t1=t2 will build the equivalence for point-set of {t1.a, t2.a} == {t1.a, t2.a}. Finally, when arriving the root aggregate node, we will build strict FD from group-by point-set {t2.a} to select aggregate function {count(t2.b)} reasonably. Consequently, we get FD as {} -> {t1.c} & {t1.a, t2.a} == {t1.a, t2.a} & {t2.a} -> {count(t2.b)}.
+
+Application:
+
+For the only-full-group-check of building phase, after getting the FD as illustrated in the last subchapter, we can easily identify the select list that t1.a are equivalent to the group-by column t2.a, count(t2.b) are strictly functional dependent on group-by column t2.a and t.c are definitely constant after the filter is passed. So only full group check is valid here. For more sophisticated cases with recursive sub-queries, maintenance work of this kind of FD architecture will be automatically and reasonable when we're just pulling the request from the top node of the entire logical plan tree, all things are naturally.
```

Comment:
```suggestion
For the only-full-group-by check of building phase, after getting the FD as illustrated in the last subchapter, we can easily identify from the select list that t1.a are equivalent to the group-by column t2.a, count(t2.b) are strictly functional dependent on group-by column t2.a and t.c are definitely constant after the filter is passed. So it can pass the only-full-group-by check here. For more sophisticated cases with recursive sub-queries, maintenance work of this kind of FD architecture will be automatical and reasonable when we're just pulling the request from the top node of the entire logical plan tree, all things happen naturally.
```